Proposal: sort set values from command line before merging

## test1 - A universal and inherited chart

values.yaml

```
...
ports: []
...
``` 

service.yaml

```
ports:
{{- range .Values.ports }}
- port: {{ .port }}
  protocol: {{ .protocol }}
{{- end }}
```

## chart test2

requirements.yaml

```
dependencies:
- name: test1
...
```

if I want to override the ports in chart test2

```
/ $: helm install --set test1.ports\[0\].port=8080 --set test1.ports\[0\].protocol=TCP --set test1.ports\[1\].port=9090 --set test1.ports\[1\].protocol=TCP ...
```
this command works as expected, but if I change the order of the  sets(In most cases, it is impossible to do so), like

```
/ $: helm install --set test1.ports\[1\].port=8080 --set test1.ports\[0\].protocol=TCP --set test1.ports\[0\].port=9090 --set test1.ports\[1\].protocol=TCP ...
```

This will lead to assertion panic

```
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 1 [running]:
k8s.io/helm/pkg/strvals.(*parser).listItem(0xc42087bf80, 0xc42087bf40, 0x2, 0x2, 0x0, 0x1, 0x5b, 0x0, 0x0, 0x0)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:291 +0xef3
k8s.io/helm/pkg/strvals.(*parser).key(0xc42087bf80, 0xc420807530, 0xc420cdd4b8, 0x3)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:177 +0xc1b
k8s.io/helm/pkg/strvals.(*parser).key(0xc42087bf80, 0xc420761f80, 0xc420cdd6f0, 0x8)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:215 +0x1a3
k8s.io/helm/pkg/strvals.(*parser).key(0xc42087bf80, 0xc420761f20, 0x1e14d7d, 0x20d7200)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:215 +0x1a3
k8s.io/helm/pkg/strvals.(*parser).parse(0xc42087bf80, 0xc420761f20, 0x0)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:133 +0x38
k8s.io/helm/pkg/strvals.ParseInto(0x7ffeefbff8c2, 0x22, 0xc420761f20, 0x0, 0x0)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:85 +0xbf
main.vals(0x0, 0x0, 0x0, 0xc420274600, 0xd, 0x10, 0x2cd7b28, 0x0, 0x0, 0x2cd7b28, ...)
	/go/src/k8s.io/helm/cmd/helm/install.go:382 +0x45a
main.(*installCmd).run(0xc4201d0600, 0x0, 0x2378600)
	/go/src/k8s.io/helm/cmd/helm/install.go:232 +0x1e4
main.newInstallCmd.func2(0xc4207666c0, 0xc420310780, 0x1, 0x1d, 0x0, 0x0)
	/go/src/k8s.io/helm/cmd/helm/install.go:191 +0x326
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).execute(0xc4207666c0, 0xc4203101e0, 0x1d, 0x1e, 0xc4207666c0, 0xc4203101e0)
	/go/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:599 +0x3db
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4207b0000, 0xc42071ab40, 0xc42071b440, 0xc42071bd40)
	/go/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:689 +0x2d4
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4207b0000, 0x1e, 0x1f)
	/go/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.main()
	/go/src/k8s.io/helm/cmd/helm/helm.go:161 +0x79
```

Sort values before invoking function `strvals.ParseInto` and `strvals.ParseIntoString` might do the trick

--- a/cmd/helm/install_test.go
+++ b/cmd/helm/install_test.go
@@ -61,7 +61,13 @@ func TestInstall(t *testing.T) {
 			resp:     helm.ReleaseMock(&helm.MockReleaseOptions{Name: "virgil"}),
 			expected: "virgil",
 		},
-		// Install, values from yaml
+		{
+			name:     "install with multiple unordered list values",
+			args:     []string{"testdata/testcharts/alpine"},
+			flags:    strings.Split("--name virgil --set foo[1].bar=baz,foo[0].baz=bar", " "),
+			resp:     helm.ReleaseMock(&helm.MockReleaseOptions{Name: "virgil"}),
+			expected: "virgil",
+		},
 		{
 			name:     "install with values",
 			args:     []string{"testdata/testcharts/alpine"},

--- a/pkg/strvals/parser.go
+++ b/pkg/strvals/parser.go
@@ -288,7 +288,13 @@ func (t *parser) listItem(list []interface{}, i int) ([]interface{}, error) {
 		// We have a nested object. Send to t.key
 		inner := map[string]interface{}{}
 		if len(list) > i {
-			inner = list[i].(map[string]interface{})
+			var ok bool
+			inner, ok = list[i].(map[string]interface{})
+			if !ok {
+				// We have indices out of order. Initialize empty value.
+				list[i] = map[string]interface{}{}
+				inner = list[i].(map[string]interface{})
+			}
 		}
 
 		// Recurse

--- a/pkg/strvals/parser_test.go
+++ b/pkg/strvals/parser_test.go
@@ -294,6 +294,30 @@ func TestParseSet(t *testing.T) {
 			str:    "nested[1][1]=1",
 			expect: map[string]interface{}{"nested": []interface{}{nil, []interface{}{nil, 1}}},
 		},
+		{
+			str: "name1.name2[0].foo=bar,name1.name2[1].foo=bar",
+			expect: map[string]interface{}{
+				"name1": map[string]interface{}{
+					"name2": []map[string]interface{}{{"foo": "bar"}, {"foo": "bar"}},
+				},
+			},
+		},
+		{
+			str: "name1.name2[1].foo=bar,name1.name2[0].foo=bar",
+			expect: map[string]interface{}{
+				"name1": map[string]interface{}{
+					"name2": []map[string]interface{}{{"foo": "bar"}, {"foo": "bar"}},
+				},
+			},
+		},
+		{
+			str: "name1.name2[1].foo=bar",
+			expect: map[string]interface{}{
+				"name1": map[string]interface{}{
+					"name2": []map[string]interface{}{nil, {"foo": "bar"}},
+				},
+			},
+		},
 	}
 
 	for _, tt := range tests {

